### PR TITLE
T210-007 Add a script which digests traces

### PR DESCRIPTION
**Pull request by setton**
For the purpose of debugging and investigating memory issues.

### CODE REVIEW

File: scripts/digest_traces.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            print("{}:{}{}\t{}\t{}".format(j[1],
+                                           " " * max(1, (40 - len(j[1]))),
+                                           self.methods_count[j[1]],
+                                           self.methods_exception_count[j[1]],
+                                           j[0]))
+        print("\nTOTAL:  {}\n".format(total))
+
+
+def main():
+    if len(sys.argv) < 2:
```
Comment (AdrienBoulanger) [inline scripts/digest_traces.py]: If it only work with one argument as the input trace file then it should be "if len(sys.argv) != 2:"

File: scripts/digest_traces.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+        # read the file
+        with open(filename, "r") as f:
+            for line in f.readlines():
+                self.read_line(line.strip())
+
+    def read_line(self, line):
+        """Read and process one line of traces"""
+        if (not self.currently_looking_at_exception and
+           line.startswith("_ALS.MAIN")):
```
Comment (AdrienBoulanger) [inline scripts/digest_traces.py]: You are assuming that any multi-lines trace in "ALS.MAIN" is an exception.

File: scripts/digest_traces.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        self.in_re = re.compile(TRACES_LINE)
+
+        self.last_method = "(elaboration)"
+        self.last_memory_usage = 0
+        # the last method and last memory usage received
+
+        self.methods_to_memory = {}
+        # cumulative memory increase indexed by method id
+
+        self.methods_exception_count = {"(elaboration)": 0}
```
Comment (AdrienBoulanger) [inline scripts/digest_traces.py]: You should replace the hard-coded "(elaboration)" by self.last_method

File: scripts/digest_traces.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        # count the exception occurrences for each exception
+
+        self.currently_looking_at_exception = False
+
+        self.methods_count = {"(elaboration)": 1}
```
Comment (AdrienBoulanger) [inline scripts/digest_traces.py]: Same as previous

File: scripts/digest_traces.py (hunk 1 above)
Comment (AdrienBoulanger) [inline scripts/digest_traces.py]: This script can only handle one argument => the condition should be len(sys.argv) != 2

File: scripts/digest_traces.py (hunk 2 above)
Comment (setton) [reply scripts/digest_traces.py]: yes, in this version we have no better way of doing this! I will modify the ALS to say "EXCEPTION" shortly.

File: scripts/digest_traces.py (hunk 3 above)
Comment (setton) [reply scripts/digest_traces.py]: done

File: scripts/digest_traces.py (hunk 4 above)
Comment (setton) [reply scripts/digest_traces.py]: done

File: scripts/digest_traces.py (hunk 1 above)
Comment (setton) [reply scripts/digest_traces.py]: done

**Pull request merged**